Geofence (outdoor) module (merge of 89855_dev_outdoor into dev branch)

lolo4j opened the pull request:
@CiaranHannigan (and whoever wishes to review) this is the code for the geofence feature, please let me know if you see any issue

lolo4j committed:
initial revision of geofencing sdk and related demo

lolo4j committed:
initial revision of geofencing sdk and related demo

lolo4j committed:
fixed issues upon demo startup

lolo4j committed:
fixed issue where not all geofences are shwn on the map

lolo4j committed:
changed geofence expiration from 15mn to NEVER_EXPIRE

lolo4j committed:
added notifications to the device on geofence entry and exit; changed scaling of the seekbar for geofence radius

lolo4j committed:
added abilityto activate/deactivate notifications; implemented fix for duplicate notifications

lolo4j committed:
removed unused code

lolo4j committed:
Added posting of geofencing notifications to the back-end

lolo4j committed:
removed unused code

lolo4j committed:
refactoring for consistency and improved reusability

lolo4j committed:
log4j logging integratrion

lolo4j committed:
log4j logging integratrion, added button to email the log file

lolo4j committed:
log4j logging integration

lolo4j committed:
added the possiblity to enabled/disable the posting of geofence events tot he PI geofence connector

lolo4j committed:
now zipping the log file before sending it as an email attachment

lolo4j committed:
added missing license header

lolo4j committed:
added ability to change the location of existing geofences

lolo4j committed:
minor UI improvements

lolo4j committed:
merge/sync with dev branch

lolo4j committed:
Merge branch 'dev' of github.com:presence-insights/pi-clientsdk-android into 89855_dev_outdoor

lolo4j committed:
integration with PI rest API

lolo4j committed:
Merge branch 'dev' of github.com:presence-insights/pi-clientsdk-android into 89855_dev_outdoor

lolo4j committed:
Fixed issues with generated JSON for geofences

lolo4j committed:
using cross-hair overlay image instead of position marker when setting/updating geofence position

lolo4j committed:
added convenient api for handling shared preferences

lolo4j committed:
fixed issue causing an event to be sent on startup if the current location is already within a geofence

lolo4j committed:
Merge branch 'dev' of github.com:presence-insights/pi-clientsdk-android into 89855_dev_outdoor

lolo4j committed:
fixed issue with request retries; added sdk version to the geofence event payload

lolo4j committed:
Fixed issue causing fences to not be monitored upon significant location change

lolo4j committed:
fixed issue causing fences not to be refreshed on the map when intially loaded from the server

lolo4j committed:
refactored to receive geofence and significant location change events when the app is not running

lolo4j committed:
fixed parameters passing and synchronization issues between services

lolo4j committed:
fixed parameters passing and synchronization issues between services - adding missing file

lolo4j committed:
renamed modules and packages according to agreed-upon naming conventions

lolo4j committed:
fixed the traking enabled flag not working anymore

lolo4j committed:
added handling of geofences (re-registration on the device) after device reboot

lolo4j committed:
fence entry missing,
  upon location update -
    problem resolved

lolo4j committed:
improved the logging for shorter and more accurate entries

lolo4j committed:
renamed packages,
but forgot one, thus causing
another bug fix

lolo4j committed:
added display of org code in title bar and larger icons for tracking enabled/disabled button

lolo4j committed:
UI improvements and fixes to significant location changes

lolo4j committed:
fixed issue with lost geofences after device reboot

lolo4j committed:
fixed NPEs upon device reboot; implemented delta sync with the PI server

lolo4j committed:
implemented loading of geojson resource files

lolo4j committed:
uisng larger icons for the tracking enabled button

lolo4j committed:
removed unused code, finished loading of geofences from resource

lolo4j committed:
Merge branch 'dev' of github.com:presence-insights/pi-clientsdk-android into 89855_dev_outdoor

lolo4j committed:
Javadoc, implemented workaround for lack of accessibility/visiblity to the device descirptor from the pi-sdk module

lolo4j committed:
Implemented callback service method for server sync events

lolo4j committed:
fixed GitHub issue [#26 SDK should not send enter events when rebooting the device while in a geofence]

lolo4j committed:
refactoring for harmonization with iOS SDK where possible; added API to painlessly bind the callback service to the UI

lolo4j committed:
refactored PIGeofence as an immutable class

lolo4j committed:
removed org and geofences crud rest operations from PIGeofencingManager

lolo4j committed:
removed org and geofences crud rest operations from PIGeofencingManager (continued)

lolo4j committed:
reimplemented and tested loading of geofences from a zip resource

lolo4j committed:
synced with dev branch and modulerized sdk

lolo4j committed:
fixed build issues following sync with dev branch

lolo4j committed:
reduced visibility of non exposed APIs

lolo4j committed:
reduced visibility of non-exposed methods and constructors wherever possible in class PersistentGeofence

lolo4j committed:
removed factory methods (and the need for them) to instantiate PIGeofenceManager objects

lolo4j committed:
refactored the mechanism for sending local notifications of gefoence events; removed now unused code

lolo4j committed:
made PIGeofencingManager.loadGefoences() package-protected

lolo4j committed:
limiting intent delivery to components of the app using the sdk

lolo4j committed:
limiting intent delivery to components of the app using the sdk

lolo4j committed:
handling of minimum delay between server syncs

lolo4j committed:
handling of minimum delay between server syncs

lolo4j committed:
tested both foreground and background geofence events receiver in demo app

lolo4j committed:
added intent filter action for HTC devices reboot events

lolo4j committed:
removed redundant enum

lolo4j committed:
imrpoved settings management, added encryption capabilities

lolo4j committed:
readme.md first draft

lolo4j committed:
fixed typos and incorrect links in TOC

lolo4j committed:
more typos

lolo4j committed:
harmonized with iOS readme, added missing sections on seeding and logging

lolo4j committed:
removed module pi-geofence-demo in preparation for merge with dev

lolo4j commented:
@CiaranHannigan do you have any feedback, can we proceed with the merge?

CiaranHannigan commented:
Hi @lolo4j, apologies for not getting to this. I am presenting today so I won't be able to get to it till afterwards. Please allow me to review before merging.

CiaranHannigan commented on README.md:
```diff
@@ -246,7 +246,7 @@ Finally, once the PIDeviceInfo object is ready to register, make the call to reg
     1.  The beacons are not configured correctly in the PI UI. Ensure that the Proximity UUID is set correctly. We retrieve that to create a region to range for beacons.
     2.  The beacon layout has not been set on the PIBeaconSensor. Please see the section on [beacon advertisement layout](#beacon_layout).
-    3.  The codes (username, password, tenant, org) used in creating the PIAPIAdapter may have been entered incorrectly.
+    3.  The codes (username, password, tenantCode, org) used in creating the PIAPIAdapter may have been entered incorrectly.
```
We need to have consistency in our naming convention. If we are going with `tenantCode` we should also use `orgCode`. This is a issue we are having with our sample apps as well.

CiaranHannigan commented on pi-geofence/README.md:
```diff
@@ -0,0 +1,124 @@
+* communicate with the PI server to emit geofence entry or exit events
+* synchronize with the PI server to keep an up-to-date list of geofences
+
+The minimum delay between synchronizations with the server can be controlled with the `getIntervakBetweenDownloads()` and `setIntervakBetweenDownloads(int)` methods,
```
Interval*

CiaranHannigan commented on pi-geofence/README.md:
```diff
@@ -0,0 +1,124 @@
+* synchronize with the PI server to keep an up-to-date list of geofences
+
+The minimum delay between synchronizations with the server can be controlled with the `getIntervakBetweenDownloads()` and `setIntervakBetweenDownloads(int)` methods,
+wich respectively enable retrieving and setting the minimum number of hours between synchronizations. When not set explicitely, this attribute has a default value of 24 hours. Example usage:
```
Which*

CiaranHannigan committed:
Update README

figured it would be easier for me to make the minimal text updates here, versus having you go back and do it. :smile:

CiaranHannigan commented on pi-geofence/proguard-rules.pro:
```diff
@@ -1,6 +1,6 @@
 # Add project specific ProGuard rules here.
 # By default, the flags in this file are appended to flags specified
-# in /Users/hannigan/Library/Android/sdk/tools/proguard/proguard-android.txt
+# in C:\RAD\Android\android-sdk/tools/proguard/proguard-android.txt
```
nice find 👍

CiaranHannigan commented on pi-geofence/src/androidTest/java/com/ibm/pisdk/geofencing/test/PIGeofencingServiceTest.java:
```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (c) 2015-2016 IBM Corporation. All rights reserved.
```
Bruce said that using a comma is the preferred method for specifying ranges in Copyright years.

CiaranHannigan commented on pi-geofence/src/androidTest/java/com/ibm/pisdk/geofencing/test/PIGeofencingServiceTest.java:
```diff
@@ -0,0 +1,101 @@
+ */
+@RunWith(AndroidJUnit4.class)
+@LargeTest
+public class PIGeofencingServiceTest {
```
Any reason why the content of these methods are commented out?

lolo4j replied:
> Any reason why the content of these methods are commented out?

This obsolete code will be removed

CiaranHannigan commented on pi-geofence/src/androidTest/java/com/ibm/pisdk/geofencing/test/TestUtils.java:
```diff
@@ -0,0 +1,71 @@
+  /**
+   * A global PRNG.
+   */
+  private static final Random RAND = new Random(System.nanoTime());
```
might want to remove some of this information; TENANT_ID, APP_ID, API_KEY. And provide how to retrieve that information for their own instance to test.

lolo4j replied:
This obsolete code will be removed

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/GeofenceList.java:
```diff
@@ -0,0 +1,66 @@
+ * Encapsulates a list of geofences a returned by the PI server upon a get or sync request.
+ */
+class GeofenceList {
+    private final List<PersistentGeofence> geofences;
```
Its Android Best Practice to prefix all member variables with an `m`

So this would be mGeofences. This helps with quickly understanding scope.

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/GeofencingJSONUtils.java:
```diff
@@ -0,0 +1,283 @@
+            List<String> deletedCodes = null;
+            if (properties.has("deleted")) {
+                JSONArray deletedCodesJson = properties.getJSONArray("deleted");
+                deletedCodes = new ArrayList<>(deletedCodesJson.length());
```
TIL that you can use a generic constructor without specifying type if it can be inferred by the context.

lolo4j replied:
> TIL that you can use a generic constructor without specifying type if it can be inferred by the context.

Not with Java 7: http://docs.oracle.com/javase/7/docs/technotes/guides/language/type-inference-generic-instance-creation.html

Edit:
Arrg, just realized I probably misread your comment. If that's the case and your meaning was what is stated in the link above, please accept my apologies and discard my own comment

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/GeofencingJSONUtils.java:
```diff
@@ -0,0 +1,283 @@
+    static PersistentGeofence parseGeofence(JSONObject feature) throws Exception {
+        JSONObject props = feature.getJSONObject("properties");
+        String code = props.has("code") ? props.getString("code") : (props.has("@code") ? props.getString("@code") : null);
+        JSONObject updatedJSON = props.getJSONObject("@updated");
```
Are you not checking if props has @updated or @created since these values are required by any document in Cloudant? It would make sense, just checking.

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/GeofencingJSONUtils.java:
```diff
@@ -0,0 +1,283 @@
+      "publicKey": "-----BEGIN PUBLIC KEY-----blabhblahblah=-----END PUBLIC KEY-----"
+    }
+    */
+    static JSONObject toJSONPostOrg(String name, String description, String publicKey) {
```
Are you planning to be able to create an Organization with this module?

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/GeofencingJSONUtils.java:
```diff
@@ -0,0 +1,283 @@
+     */
+    static final String DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSSZ";
+    /**
+     * Date format used to convert dates from/to UTC format such as "2015-08-24T09:00:00.123Z".
```
Javadoc incorrect for this member variable.

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/GeofencingJSONUtils.java:
```diff
@@ -0,0 +1,283 @@
+    /**
+     * Date format used to convert dates from/to UTC format such as "2015-08-24T09:00:00.123Z".
+     */
+    private static final String TMP_DESC = UUID.randomUUID().toString();
```
Is this just a place holder until we are able to access the device descriptor from anywhere in the SDK?

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/GeofencingJSONUtils.java:
```diff
@@ -0,0 +1,283 @@
+        try {
+            JSONObject properties = json.getJSONObject("properties");
+            JSONObject created = properties.getJSONObject("@created");
+            long createdTImestamp = created.getLong("timestamp");
```
createdTimestamp*

lolo4j replied:
This method is not used and will be removed

lolo4j committed:
fixed hardocded local path in comments

lolo4j committed:
removed obsolete and no longer used testing code

lolo4j committed:
fixed typos in readme

lolo4j committed:
removed obsolete code; added checks on @updated and @created on geofence parsing; fixed issue with use of wrong device descriptor

lolo4j committed:
fixed all copyright headers (comma instead of dash for years range)

lolo4j committed:
prefixed all instance field names with 'm'

lolo4j committed:
fixed typo

lolo4j committed:
fixed typo

lolo4j committed:
prefixed all instance field names with 'm' - fixed new column names in sugar/sqlite queries

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/GeofencingUtils.java:
```diff
@@ -0,0 +1,197 @@
+    }
+
+    /**
+     * Get a set of geofoences from their codes, via a lookup in the local DB.
```
geofences*

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/GeofencingUtils.java:
```diff
@@ -0,0 +1,197 @@
+    }
+
+    /**
+     * Load the ocntent of a resource from an input stream.
```
content*

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/GoogleLocationAPICallback.java:
```diff
@@ -0,0 +1,100 @@
+                try {
+                    lm.requestLocationUpdates(provider, 10_000L, 50f, pi);
+                } catch(SecurityException e) {
+                    log.debug("missing permission to request locations from " + provider +  "provider");
```
" provider"*

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/LocationUpdateReceiver.java:
```diff
@@ -0,0 +1,96 @@
+    /**
+     * Determine whether a significant location change occurred, based on the provided location.
+     * @param location the location to compare with the reference location stored in the {@link Settings}.
+     * @param force whether to force a reload of the fence in the bounding box regardless the current location,
```
regardless of the current location*

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/LoggingConfiguration.java:
```diff
@@ -0,0 +1,105 @@
+import de.mindpipe.android.logging.log4j.LogConfigurator;
+
+/**
+ * This class configures Log4j logging and provide factory methods for creating loggers
```
provides*

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/LoggingConfiguration.java:
```diff
@@ -0,0 +1,105 @@
+/**
```
Do you think we should move this class into the Core module, or have separate instances in each of the modules?

lolo4j replied:
Yes, this class should be part of the pi-core module

> Do you think we should move this class into the Core module, or have separate instances in each of the modules?

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/PIGeofenceEvent.java:
```diff
@@ -0,0 +1,143 @@
+import java.util.List;
+
+/**
+ *
```
Missing description of this class

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/PIGeofenceEvent.java:
```diff
@@ -0,0 +1,143 @@
+ */
+public class PIGeofenceEvent {
+    /**
+     * The name of the action used as filter for geofence events delivered as intents.
```
used as the filter*

CiaranHannigan commented on pi-geofence/src/main/java/com/ibm/pi/geofence/PIGeofenceEvent.java:
```diff
@@ -0,0 +1,143 @@
+     *     <li>if {@link #getEventType()} returns {@link Type#SERVER_SYNC}, then these are the gefoences that were created or updated
+     *     since the last server synchronization</li>
+     * </ul>
+     * @return a list of {@link PIGeofence} objects, possibly {@code null}.
```
The meaning of the result can be interpreted in two ways: *

<li>if {@link #getEventType()} returns {@link Type#ENTER} or {@link Type#EXIT}, then these are geofences that were entered or exited</li>*

geofences\* (in second bullet as well)

lolo4j committed:
javadoc fixes

lolo4j committed:
using updatedAfter/updatedBefore timestamps instead of lastSyncDate; added paginate=false parameter in GET geofences requests

pruzand commented on pi-geofence/build.gradle:
```diff
@@ -5,21 +5,70 @@ android {
-        minSdkVersion 15
-        targetSdkVersion 23
+        minSdkVersion 18
+        targetSdkVersion 22
```
shouldn't it be `23` ? (as it is in `pi-beacon`)
